### PR TITLE
PB-1934: Fetching backup size from backup response JSON content

### PR DESCRIPTION
**Pull request by prashanthpx**
**What this PR does / why we need it**:
Fetching backup size from backup response JSON content

**Which issue(s) this PR fixes** (optional)
PB-1934

**Special notes for your reviewer**:

Manual test
DataExprot CR once the backup is successful
```
[root@prashanth-iron-seer-0 kdmp]# k describe dataexport kopia-aws-mysql-1 -nrst
Name:         kopia-aws-mysql-1
Namespace:    rst
Labels:       <none>
Annotations:  <none>
API Version:  kdmp.portworx.com/v1alpha1
Kind:         DataExport
Metadata:
  Creation Timestamp:  2021-09-29T13:17:33Z
  Finalizers:
    kdmp.portworx.com/finalizer-cleanup
  Generation:  12
  Managed Fields:
    API Version:  kdmp.portworx.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:destination:
 Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-09-29T13:17:33Z
    API Version:  kdmp.portworx.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
      f:status:
        .:
        f:size:
        f:snapshotID:
        f:stage:
        f:status:
        f:transferID:
    Manager:         kdmp
    Operation:       Update
    Time:            2021-09-29T13:18:03Z
  Resource Version:  33561246
  UID:               5d26c541-f089-47fd-a396-e9bd5752054e
Spec:
  Destination:
    API Version:  stork.libopenstorage.org/v1alpha1
    Kind:         BackupLocation
    Name:         kopia-bl
    Namespace:    rst
  Source:
    API Version:  v1
    Kind:         PersistentVolumeClaim
    Name:         mysql-data
    Namespace:    rst
  Type:           kopia
Status:
  Size:         120861096
  Snapshot ID:  bac001d193125b51883ae3d7c6e26ea1
  Stage:        Final
  Status:       Successful
  Transfer ID:  rst/b-kopia-aws-mysql-1
Events:         <none>
```

**Note**: Corresponding changes in stork needs to be PR to populate the size backup in ApplicationBackup CR

### CODE REVIEW

File: pkg/kopia/backup.go [hunk 1]
```diff
@@ -182,5 +183,29 @@ func getBackupSummary(outBytes []byte, errBytes []byte) (*BackupSummaryResponse,
 			CmdErr:    string(errBytes),
 		}
 	}
+	// TODO: There is an issue in parsing json output for embedded size structure.
+	// For now going ahead with string parsing to fetch the size.
+	// Eg JSOn response
```
Comment (adityadani) [inline pkg/kopia/backup.go:188]: Can you elaborate more on the issue here? What was the object you defined to unmarshal this string into ?

Comment (prashanthpx) [reply pkg/kopia/backup.go:188]: Currently, the response structure for backup response is as follows
```
// BackupSummaryResponse describes single snapshot entry.
type BackupSummaryResponse struct {
	ID               string           `json:"id"`
	Source           SourceInfo       `json:"source"`
	Description      string           `json:"description"`
	StartTime        time.Time        `json:"startTime"`
	EndTime          time.Time        `json:"endTime"`
	IncompleteReason string           `json:"incomplete,omitempty"`
	Summary          DirectorySummary `json:"summ"`
	RootEntry        string           `json:"rootID"`
	RetentionReasons []string         `json:"retention"`
}


// SourceInfo represents the information about snapshot source.
type SourceInfo struct {
	Host     string `json:"host"`
	UserName string `json:"userName"`
	Path     string `json:"path"`
}

// DirectorySummary represents summary information about a directory.
type DirectorySummary struct {
	TotalFileSize     uint64    `json:"size"`
	TotalFileCount    int64     `json:"files"`
	TotalSymlinkCount int64     `json:"symlinks"`
	TotalDirCount     int64     `json:"dirs"`
	MaxModTime        time.Time `json:"maxTime"`
	IncompleteReason  string    `json:"incomplete,omitempty"`

	// number of failed files
	FatalErrorCount   int `json:"numFailed"`
	IgnoredErrorCount int `json:"numIgnoredErrors,omitempty"`

	FailedEntries []*EntryWithError `json:"errors,omitempty"`
}

```

Here if we see the size is part of `DirectorySummary` structure which is used in `BackupSummaryResponse` structure.
The issue was when unmarshalled, the `DirectorySummary` is not getting unmarshalled and values are as zero. All other data like ID, Source, etc.. are getting unmarshalled

Below is sample of output response when printed
```

outResponse: {\"id\":\"fec958126a2e11e038b67789fb7ea21c\",\"source\":{
\"host\":\"b-kopia-aws-mysql-1-q9whf\",\"userName\":\"root\",\"path\":\"/data/kubernetes.io~portworx-volume/pvc-c6826179-ac
f4-481d-a724-b1b6ff5f444c\"},\"description\":\"\",\"startTime\":\"2021-09-29T12:12:11.819642442Z\",\"endTime\":\"2021-09-29
T12:12:12.40206701Z\",\"rootEntry\":{\"name\":\"pvc-c6826179-acf4-481d-a724-b1b6ff5f444c\",\"type\":\"d\",\"mode\":\"0755\"
,\"mtime\":\"2021-07-09T11:09:33.641023675Z\",\"uid\":999,\"obj\":\"k6db0cc298b1a4251799a519f7941b134\",\"summ\":{\"size\":
120861096,\"files\":136,\"symlinks\":0,\"dirs\":3,\"maxTime\":\"2021-07-09T11:09:35.734024033Z\",\"numFailed\":0}}}"
```
Here we can see the `summ` is the JSON tag for summary response which holds size info and other information.

I tried unmarshalling using the following ways
1. json.Unmarshal(outResponse, summaryResponse);

2. yaml.NewYAMLOrJSONDecoder()

Both of them don't give size value.
I had also posted the question in kopia forum, haven't yet got any response.

Comment (adityadani) [reply pkg/kopia/backup.go:188]: @prashanthpx I think you have defined the summary and rootEntry objects incorrectly. I used the above `outputResponse` and wrote this program and was successfully able to unmarshal the summary object

https://play.golang.org/p/NZurO2iIv9e

You should define the structs like this:

```
// BackupSummaryResponse describes single snapshot entry.
type BackupSummaryResponse struct {
	ID               string           `json:"id"`
	Source           SourceInfo       `json:"source"`
	Description      string           `json:"description"`
	StartTime        time.Time        `json:"startTime"`
	EndTime          time.Time        `json:"endTime"`
	IncompleteReason string           `json:"incomplete,omitempty"`
	RootEntry        RootEntry           `json:"rootEntry"`
	RetentionReasons []string         `json:"retention"`
}

type RootEntry struct {
	Name string `json:"name"`
	Type string `json:"type"`
	Mode string `json:"mode"`
	ModifiedTime time.Time `json:"mtime"`
	UID uint64 `json:"uid"`
	ObjectIdentifier string `json:"obj"`
	Summary          DirectorySummary `json:"summ"`
}


// SourceInfo represents the information about snapshot source.
type SourceInfo struct {
	Host     string `json:"host"`
	UserName string `json:"userName"`
	Path     string `json:"path"`	
}

// DirectorySummary represents summary information about a directory.
type DirectorySummary struct {
	TotalFileSize     uint64    `json:"size"`
	TotalFileCount    int64     `json:"files"`
	TotalSymlinkCount int64     `json:"symlinks"`
	TotalDirCount     int64     `json:"dirs"`
	MaxModTime        time.Time `json:"maxTime"`
	IncompleteReason  string    `json:"incomplete,omitempty"`

	// number of failed files
	FatalErrorCount   int `json:"numFailed"`
	IgnoredErrorCount int `json:"numIgnoredErrors,omitempty"`

	//FailedEntries []*EntryWithError `json:"errors,omitempty"`
}
```

Comment (prashanthpx) [reply pkg/kopia/backup.go]: Thanks @adityadani for the help. I had actually taken the structure from the kopia repository and looks like their output JSON format doesn't match the structure. I made the changes and works now.  Can you please review the same?

**Review by adityadani — APPROVED** [review]

**Pull request merged**